File: 백준/Silver/16948. 데스 나이트/데스 나이트.py
import sys
input = sys.stdin.readline
from collections import deque

n = int(input().rstrip())
r1,c1,r2,c2 = map(int,input().split())
flag = ''
board = [[0]*n for _ in range(n)]

dx = [-2,-2,0,0,2,2]
dy = [-1,1,-2,2,-1,1]
queue = deque([])

def bfs(i,j):
    queue.append([i,j])

    while queue:
        x,y = queue.popleft()
        if x == r2 and y == c2:
            return True
        for k in range(6):
            nx = x+dx[k]
            ny = y+dy[k]
            if 0<=nx<n and 0<=ny<n and board[nx][ny] == 0:
                board[nx][ny] = board[x][y] + 1
                queue.append([nx,ny])

    return False

flag = bfs(r1,c1)

if flag:
    print(board[r2][c2])
else:
    print(-1)

File: 백준/Silver/16948. 데스 나이트/데스 나이트.js
const fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

let N = Number(input.shift());
let board = Array.from(Array(N),() => Array(N).fill(0));
let answer = '';

// 시작(bfs 걸기) bfs내에 좌표 r2,c2 만나면 종료
let [r1,c1,r2,c2] = input[0].split(' ').map(Number);

let dx = [-2,-2,0,0,2,2];
let dy = [-1,1,-2,2,-1,1];
let queue = [];

function bfs(i,j) {
    queue.push([i,j])

    while(queue.length){
        let [x,y] = queue.shift();
        if (x===r2 && y === c2) {
            return true           
        }
        for(let k=0; k<6; k++){
            let nx = x+dx[k]
            let ny = y+dy[k]
            if(0<=nx && nx<N && 0<=ny && ny<N && board[nx][ny]===0){
                board[nx][ny] = board[x][y] + 1;
                queue.push([nx,ny]);
            }
        }
    }
    return false
}

answer = bfs(r1,c1);
answer === true ? console.log(board[r2][c2]) : console.log(-1);